File: public/python/029-copy-list-with-random-pointer.py
# LeetCode Problem #138: Copy List with Random Pointer
# Difficulty: Medium
# Link: https://leetcode.com/problems/copy-list-with-random-pointer/

from typing import Optional


class Node:
    def __init__(self, x: int, next=None, random=None):
        self.val = int(x)
        self.next = next
        self.random = random


# ─────────────────────────────────────────────
# APPROACH 1: Hash Map (Intuitive) | O(n) time | O(n) space
# EXPLAIN: Two passes: first create all copied nodes stored in a map; second wire next and random pointers via the map.
# WHEN: Most readable approach; standard first solution in an interview.

def copyRandomList_hashmap(head: Optional[Node]) -> Optional[Node]:
    if not head:
        return None
    old_to_new = {}
    curr = head
    while curr:
        old_to_new[curr] = Node(curr.val)
        curr = curr.next
    curr = head
    while curr:
        if curr.next:
            old_to_new[curr].next = old_to_new[curr.next]
        if curr.random:
            old_to_new[curr].random = old_to_new[curr.random]
        curr = curr.next
    return old_to_new[head]


# ─────────────────────────────────────────────
# APPROACH 2: Interweaving Nodes (Optimal) | O(n) time | O(1) space
# EXPLAIN: Insert each copy right after its original (A->A'->B->B'...), set random pointers, then separate the lists.
# WHEN: Best space efficiency — O(1) extra, no hash map needed.

def copyRandomList_interweave(head: Optional[Node]) -> Optional[Node]:
    if not head:
        return None
    # Step 1: Interweave
    curr = head
    while curr:
        copy = Node(curr.val)
        copy.next = curr.next
        curr.next = copy
        curr = copy.next
    # Step 2: Set random pointers
    curr = head
    while curr:
        if curr.random:
            curr.next.random = curr.random.next
        curr = curr.next.next
    # Step 3: Separate lists
    dummy = Node(0)
    copy_curr = dummy
    curr = head
    while curr:
        copy_curr.next = curr.next
        curr.next = curr.next.next
        copy_curr = copy_curr.next
        curr = curr.next
    return dummy.next


# ─────────────────────────────────────────────
# APPROACH 3: Recursive with Hash Map | O(n) time | O(n) space
# EXPLAIN: Recursively clone each node, using a visited dict to avoid creating duplicates.
# WHEN: Natural recursive formulation; useful when you think about the problem structurally.

def copyRandomList_recursive(head: Optional[Node]) -> Optional[Node]:
    visited: dict = {}

    def clone(node):
        if not node:
            return None
        if node in visited:
            return visited[node]
        copy = Node(node.val)
        visited[node] = copy
        copy.next = clone(node.next)
        copy.random = clone(node.random)
        return copy

    return clone(head)


# ─────────────────────────────────────────────
# APPROACH 4: Hash Map with Null Sentinel | O(n) time | O(n) space
# EXPLAIN: Pre-insert None->None into the map so next/random pointer assignments need no null checks.
# WHEN: Slightly cleaner code than Approach 1 by handling null uniformly.

def copyRandomList_null_sentinel(head: Optional[Node]) -> Optional[Node]:
    if not head:
        return None
    mp: dict = {None: None}
    curr = head
    while curr:
        mp[curr] = Node(curr.val)
        curr = curr.next
    curr = head
    while curr:
        mp[curr].next = mp[curr.next]
        mp[curr].random = mp[curr.random]
        curr = curr.next
    return mp[head]


# ─────────────────────────────────────────────
# APPROACH 5: Vector/Index-Based | O(n) time | O(n) space
# EXPLAIN: Store nodes in a list, build an index map, create copies, then wire pointers using indices.
# WHEN: Useful when you want to think about nodes by position rather than pointer identity.

def copyRandomList_index(head: Optional[Node]) -> Optional[Node]:
    if not head:
        return None
    nodes = []
    node_to_idx: dict = {}
    curr = head
    while curr:
        node_to_idx[curr] = len(nodes)
        nodes.append(curr)
        curr = curr.next
    copies = [Node(n.val) for n in nodes]
    for i, n in enumerate(nodes):
        if n.next:
            copies[i].next = copies[i + 1]
        if n.random:
            copies[i].random = copies[node_to_idx[n.random]]
    return copies[0]


# Made with Bob

File: public/java/029-copy-list-with-random-pointer.java
/*
 * LeetCode Problem #138: Copy List with Random Pointer
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/copy-list-with-random-pointer/
 */
import java.util.*;

public class Solution {

    static class Node {
        int val;
        Node next, random;
        Node(int val) { this.val = val; }
    }

    // APPROACH 1: Hash Map | O(n) time | O(n) space
    // EXPLAIN: First pass creates all copied nodes in a map; second pass wires next and random pointers using the map.
    public Node copyRandomList(Node head) {
        if (head == null) return null;
        Map<Node, Node> map = new HashMap<>();
        for (Node c = head; c != null; c = c.next) map.put(c, new Node(c.val));
        for (Node c = head; c != null; c = c.next) {
            map.get(c).next   = map.get(c.next);
            map.get(c).random = map.get(c.random);
        }
        return map.get(head);
    }

    // APPROACH 2: Interweaving Nodes | O(n) time | O(1) space
    // EXPLAIN: Insert A'->A->B'->B...; set random pointers (A'.random = A.random.next); then separate the two lists.
    public Node copyRandomListInterweave(Node head) {
        if (head == null) return null;
        // Step 1: Interweave
        Node curr = head;
        while (curr != null) {
            Node copy = new Node(curr.val);
            copy.next = curr.next;
            curr.next = copy;
            curr = copy.next;
        }
        // Step 2: Set random
        curr = head;
        while (curr != null) {
            if (curr.random != null) curr.next.random = curr.random.next;
            curr = curr.next.next;
        }
        // Step 3: Separate
        Node dummy = new Node(0), copyCurr = dummy;
        curr = head;
        while (curr != null) {
            copyCurr.next = curr.next;
            curr.next = curr.next.next;
            copyCurr = copyCurr.next;
            curr = curr.next;
        }
        return dummy.next;
    }

    // APPROACH 3: Recursive with Memo | O(n) time | O(n) space
    // EXPLAIN: Recursively clone each node, storing the result in a visited map to avoid creating duplicates when random creates back-references.
    private Map<Node, Node> memo = new HashMap<>();

    public Node copyRandomListRecursive(Node node) {
        if (node == null) return null;
        if (memo.containsKey(node)) return memo.get(node);
        Node copy = new Node(node.val);
        memo.put(node, copy);
        copy.next   = copyRandomListRecursive(node.next);
        copy.random = copyRandomListRecursive(node.random);
        return copy;
    }

    // APPROACH 4: Hash Map with Null Sentinel | O(n) time | O(n) space
    // EXPLAIN: Pre-insert null->null so all pointer assignments are uniform without null-checks.
    public Node copyRandomListNullSentinel(Node head) {
        if (head == null) return null;
        Map<Node, Node> map = new HashMap<>();
        map.put(null, null);
        for (Node c = head; c != null; c = c.next) map.put(c, new Node(c.val));
        for (Node c = head; c != null; c = c.next) {
            map.get(c).next   = map.get(c.next);
            map.get(c).random = map.get(c.random);
        }
        return map.get(head);
    }

    // APPROACH 5: Index-Based | O(n) time | O(n) space
    // EXPLAIN: Store nodes in an array and build an index map; use indices to connect copies' random pointers.
    public Node copyRandomListIndex(Node head) {
        if (head == null) return null;
        List<Node> nodes = new ArrayList<>();
        Map<Node, Integer> idx = new HashMap<>();
        for (Node c = head; c != null; c = c.next) { idx.put(c, nodes.size()); nodes.add(c); }
        Node[] copies = new Node[nodes.size()];
        for (int i = 0; i < nodes.size(); i++) copies[i] = new Node(nodes.get(i).val);
        for (int i = 0; i < nodes.size(); i++) {
            Node orig = nodes.get(i);
            if (orig.next   != null) copies[i].next   = copies[i + 1];
            if (orig.random != null) copies[i].random = copies[idx.get(orig.random)];
        }
        return copies[0];
    }
}

// Made with Bob
